clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitResponse.java

```diff
@@ -119,43 +130,62 @@ public boolean shouldClientThrottle(short version) {
         return version >= 4;
     }
 
-    public static class Builder {
-        OffsetCommitResponseData data = new OffsetCommitResponseData();
-        HashMap<String, OffsetCommitResponseTopic> byTopicName = new HashMap<>();
+    public short version() {
+        return version;
+    }
 
-        private OffsetCommitResponseTopic getOrCreateTopic(
-            String topicName
-        ) {
-            OffsetCommitResponseTopic topic = byTopicName.get(topicName);
-            if (topic == null) {
-                topic = new OffsetCommitResponseTopic().setName(topicName);
-                data.topics().add(topic);
-                byTopicName.put(topicName, topic);
-            }
-            return topic;
+    public static Builder<?> newBuilder(TopicResolver topicResolver, short version) {
+        if (version >= 9) {
+            return new Builder<>(topicResolver, new ByTopicId(), version);
+        } else {
+            return new Builder<>(topicResolver, new ByTopicName(), version);
         }
+    }
 
-        public Builder addPartition(
-            String topicName,
-            int partitionIndex,
-            Errors error
-        ) {
-            final OffsetCommitResponseTopic topicResponse = getOrCreateTopic(topicName);
+    public static final class Builder<T> {
+        private final TopicResolver topicResolver;
+        private final TopicClassifier<T> topicClassifier;
+        private final short version;
 
-            topicResponse.partitions().add(new OffsetCommitResponsePartition()
-                .setPartitionIndex(partitionIndex)
-                .setErrorCode(error.code()));
+        private OffsetCommitResponseData data = new OffsetCommitResponseData();
+        private final Map<T, OffsetCommitResponseTopic> topics = new HashMap<>();
+
+        protected Builder(TopicResolver topicResolver, TopicClassifier<T> topicClassifier, short version) {
+            this.topicResolver = topicResolver;
+            this.topicClassifier = topicClassifier;
+            this.version = version;
+        }
+
+        public Builder<T> addPartition(String topicName, Uuid topicId, int partitionIndex, Errors error) {
+            Uuid resolvedTopicId = maybeResolveTopicId(topicName, topicId);
+
+            if (version >= 9 && Uuid.ZERO_UUID.equals(resolvedTopicId)) {
+                Errors reported = error != Errors.NONE ? error : Errors.UNKNOWN_TOPIC_ID;
```

Comment:
This case shouldn't be reachable because once we have proceeded with constructing the response via `addPartition` all topic ids are supposed to have been resolved successfully. Here, we choose to add the topic to the response with the error code `UNKNOWN_TOPIC_ID` if no error is already set. Any existing error is not overwritten.